rest-api-backend/main.go

```diff
@@ -0,0 +1,177 @@
+package main
+
+import (
+	"database/sql"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	_ "github.com/go-sql-driver/mysql"
+	"github.com/gorilla/mux"
+)
+
+// Users struct contains 2 properies that represents all the users on the site
+type Users struct {
+	Userid   string `json:"userid"`
+	Username string `json:"username"`
+}
+
+var db *sql.DB
+var err error
+
+func main() {
+
+	// connecting to the database
+	db, err = sql.Open("mysql", "root:@/lottery")
+
+	if err != nil {
+		panic(err.Error())
+	}
+
+	defer db.Close()
+
+	// creating new gorilla/mux router
+	router := mux.NewRouter()
+
+	router.HandleFunc("/", homePage)
+	router.HandleFunc("/users", getUsers).Methods("GET")
+	router.HandleFunc("/users", createUser).Methods("POST")
+	router.HandleFunc("/users/{userid}", getUser).Methods("GET")
+	router.HandleFunc("/users/{userid}", updateUser).Methods("PUT")
+	router.HandleFunc("/users/{userid}", deleteUser).Methods("DELETE")
+
+	// creating connection to local browser
+	http.ListenAndServe(":30000", router)
+}
+
+func homePage(w http.ResponseWriter, r *http.Request) {
+	fmt.Fprintf(w, "Welcome to the Homepage!")
+	fmt.Println("Endpoint Hit: homepage")
+}
+
+// retrieves all users from the users table and response with JSON formatted variables
+func getUsers(w http.ResponseWriter, r *http.Request) {
+
+	// sets content-type to applications/json
+	w.Header().Set("Content-Type", "application/json")
+
+	var users []Users
+
+	// query returns the rows of the database
+	result, err := db.Query("SELECT * FROM `users`")
+	if err != nil {
+		panic(err.Error())
+	}
+
+	defer result.Close()
+
+	// loop over result and for every iteration, creates new User instance
+	// then scans the result into user object
+	for result.Next() {
+		var user Users
+		err := result.Scan(&user.Userid, &user.Username)
+		if err != nil {
+			panic(err.Error())
+		}
+		users = append(users, user)
+	}
+
+	// encodes users array into JSON string
+	json.NewEncoder(w).Encode(users)
+	fmt.Println("Endpoint Hit: getUsers")
+}
+
+func createUser(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+
+	stmt, err := db.Prepare("INSERT INTO `users`(username) VALUES(?)")
+	if err != nil {
+		panic(err.Error())
+	}
+
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		panic(err.Error())
+	}
+
+	keyVal := make(map[string]string)
+	json.Unmarshal(body, &keyVal)
+	username := keyVal["username"]
+
+	_, err = stmt.Exec(username)
+	if err != nil {
+		panic(err.Error())
+	}
+
+	fmt.Fprintf(w, "New post was created")
+	fmt.Println("Endpoint Hit: createUser")
+}
+
+func getUser(w http.ResponseWriter, r *http.Request) {
```

Comment:
This might be a lot all at once, though consider this pattern:

First we create a error handler, this will handle errors when it receives them by calling `http.Error`. [See `http.Error` docs in the standard library.](https://golang.org/pkg/net/http/#Error). Which will return something to the user though the ResponseWriter `w`.
```
type errorHandler func(http.ResponseWriter, *http.Request) error

func (fn errorHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
    if err := fn(w, r); err != nil {
        http.Error(w, err.Error(), 500)
    }
}
```

This can then be used to "wrap" a function that has the form:

```
func getUser(w http.ResponseWriter, r *http.Request) error {
```

By doing:

```
router.Handle("/users/{userid}", errorHandler(getUser)).Methods("GET")
```

Note the change from `Handle` to `HandleFunc`. [See the standard library for more information.](https://godoc.org/net/http#Handler)

[See this page for the reasons why to use `Handle`, and how `errorHandler(getUser)` results in something that satisfies  the `http.Handler` interface.](https://astaxie.gitbooks.io/build-web-application-with-golang/content/en/03.4.html)

Finally, `getUser` can be refactored to something like:

```
func getUser(w http.ResponseWriter, r *http.Request) error {
	w.Header().Set("Content-Type", "application/json")
	params := mux.Vars(r)

	result, err := db.Query("SELECT userid, username FROM `users` WHERE userid = ?", params["userid"])
	if err != nil {
		return err
	}

	defer result.Close()

	var user Users

	for result.Next() {
		err := result.Scan(&user.Userid, &user.Username)
		if err != nil {
			return err
		}
	}

	json.NewEncoder(w).Encode(user)
	fmt.Println("Endpoint Hit: getUser")
}
```

Here is the reference that I found: 

https://astaxie.gitbooks.io/build-web-application-with-golang/content/en/11.1.html

Most of that book actually seems pretty good, though I don't agree with all of it. There is quite a lot here, and it will probably be a good reference.

https://astaxie.gitbooks.io/build-web-application-with-golang/content/en/03.0.html

Reply:
I disagree with my previous post. 

I believe this is a much clearer way to do this:

```
type ErrHandlerFunc  func(http.ResponseWriter, *http.Request) error
```
```
func errHandler(fn ErrHandlerFunc) http.HandlerFunc {
    return func(http.ResponseWriter, *http.Request) {
        if err := fn(w,r); err != nil {
            http.Error(w, err.Error(), 500)
        }
    }
}

And then go back to using `HandlerFunc`.

```
router.HandleFunc("/users/{userid}", errorHandler(getUser)).Methods("GET")
```

Then you can still write `getUser` with as such:

```
func getUser(w http.ResponseWriter, r *http.Request) error {
```
as that is of type
```
type ErrHandlerFunc  func(http.ResponseWriter, *http.Request) error
```

I'm curious what you think of these two different approaches.

Reply:
The second approach uses a "closure" if you want to read more about them.